discovery/service.go

```diff
@@ -0,0 +1,75 @@
+package discovery
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"golang.org/x/sync/syncmap"
+)
+
+// DefaultCacheInvalidationTimeout for updating instances
+const (
+	DefaultCacheInvalidationTimeout = 10
+)
+
+// Services for discovery service
+type Services struct {
+	ConsulClient             *api.Client
+	Instances                *syncmap.Map
+	CacheInvalidationTimeout int64
+}
+
+// GetEndpoint by service name
+func (s *Services) GetEndpoint(serviceName string) (url *url.URL, err error) {
+	resolver := s.UpdateInstences(serviceName)
+	if resolver != nil && len(resolver.endpoints) > 0 {
+		url = resolver.getHealthyInstanceEndpoint()
+	} else {
+		err = fmt.Errorf("no registered or healtly instances for service: %s", serviceName)
+	}
+	return
+}
+
+// UpdateInstences get service instances from service discovery
+func (s *Services) UpdateInstences(serviceName string) (resolver *Resolver) {
+	value, instancesExists := s.Instances.Load(serviceName)
+	if instancesExists {
+		resolver = value.(*Resolver)
+	} else {
+		s.Instances = new(syncmap.Map)
+		resolver = NewResolver(s.ConsulClient)
+		resolver.updateLastTimestamp()
+	}
+
+	if !instancesExists || time.Now().Unix()-resolver.CacheTTL >= s.CacheInvalidationTimeout {
+		if instancesExists {
+			if !resolver.tryLock() {
+				return resolver
+			}
+		} else {
+			resolver.lock.Lock()
+		}
+		entries := resolver.GetNodesFromConsul(serviceName)
+
+		resolver.endpoints = make([]*url.URL, 0)
+		for _, entry := range entries {
+			resolver.endpoints = append(resolver.endpoints, serviceEntryToURL(entry))
+		}
+		resolver.updateLastTimestamp()
```

Comment:
All this code seems quite weird for me. We get a value from a resolver to put a value in a resolver using a method from a resolver. Why all this can't happen in the resolver itself?